File: frontend/src/app/_models/Task.ts
import {Question} from "./Question";

export class Task {
  _id: object;
  private _assignees: object[];
  private _deadline: Date;
  private _name: string;
  private _description: string;
  private _questions: Question[];
  private _submittedBy: string[];

  constructor(id: object, taskAssignees: object[], taskDeadline: Date, taskName: string, taskDescription: string, taskQuestions: Question[], submittedBy: string[]) {
    this._id = id;
    this._assignees = taskAssignees;
    this._deadline = taskDeadline;
    this._name = taskName;
    this._description = taskDescription;
    this._questions = taskQuestions;
    this._submittedBy = submittedBy;
  }

  get deadline(): Date {
    // @ts-ignore
    return this._deadline;
  }

  get name(): String {
    return this._name;
  }

  get description(): String {
    return this._description;
  }

  get questions(): Question[] {
    return this._questions;
  }

  get submittedBy(): string[] {
    return this._submittedBy;
  }

  get assignees(): object[] {
    return this._assignees;
  }

  static fromJson(json){
    let questions: Question[] = [];
    for(let q of json.questions) {
      questions.push(new Question(q.name, q.type, q.options))
    }
    return new Task(json._id, json.assignees, new Date(Date.parse(json.deadline)), json.name, json.description, questions, json.submittedBy)
  }

}

File: backend/src/main/java/it/hva/dmci/ict/ewa/spaghetteria4/backend/models/Task.java
package it.hva.dmci.ict.ewa.spaghetteria4.backend.models;

import com.fasterxml.jackson.annotation.JsonProperty;
import org.springframework.data.mongodb.core.mapping.Document;

import javax.persistence.Entity;
import javax.persistence.Id;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Date;

@Document("task")
public class Task {
    @JsonProperty @Id
    private String _id;
    @JsonProperty private String name;
    @JsonProperty private String description;
    @JsonProperty private Date deadline;
    @JsonProperty private ArrayList<String> assignees;
    @JsonProperty private ArrayList<Question> questions;
    @JsonProperty private ArrayList<String> submittedBy;

    public static Task fromDto(TaskDto t) {
        return new Task(t.getName(), t.getDescription(), t.getDeadline(), t.getAssignees(), t.getQuestions(), t.getSubmittedBy());
    }

    public static Task fromRecurringTask(RecurringTask t, Date deadline) {
        return new Task(t.getName(),t.getDescription(),deadline,t.getAssignees(),t.getQuestions(), new ArrayList<>());
    }

    public Task(String name, String description, Date deadline, ArrayList<String> assignees, ArrayList<Question> questions, ArrayList<String> submittedBy) {
        this.name = name;
        this.description = description;
        this.deadline = deadline;
        this.assignees = assignees;
        this.questions = questions;
        this.submittedBy = submittedBy;
    }

    public ArrayList<String> getAssignees() {
        return assignees;
    }

    public boolean addAssignee(String branch) {
        if(assignees.contains(branch)) return false;
        assignees.add(branch);
        return true;
    }

    public boolean addSubmitted(String branch) {
        if(!assignees.contains(branch) || submittedBy.contains(branch)) return false;
        submittedBy.add(branch);
        return true;
    }

    public boolean removeSubmitted(String branch) {
        if(!assignees.contains(branch) || !submittedBy.contains(branch)) return false;
        submittedBy.remove(branch);
        return true;
    }

    public String get_id() {
        return _id;
    }

    public Date getDeadline() {
        return deadline;
    }

    public int getQuestionsLength() { return questions.size(); }

    public String getName() {
        return name;
    }

    public String getDescription() {
        return description;
    }

    public ArrayList<Question> getQuestions() {
        return questions;
    }

    public ArrayList<String> getSubmittedBy() {
        return submittedBy;
    }

    @Override
    public String toString() {
        return "Task{" +
                "_id='" + _id + '\'' +
                ", name='" + name + '\'' +
                ", description='" + description + '\'' +
                ", deadline=" + deadline +
                ", assignees=" + assignees +
                ", questions=" + questions +
                ", submittedBy=" + submittedBy +
                '}';
    }

    public void set_id(String _id) {
        this._id = _id;
    }
}
